Student Name: Wang minda

Student ID: 58521778

# JC2066 Assignment 1: College Admissions Algorithms

### Learning objectives:
- Explore algorithmic decision-making and representations of people in data
- Understanding outliers, missing information in data, and human judgments in algorithmic decisions

### About Submission
The questions include two parts: **coding** (Q1 to Q5) and **reflection** (Q6 & Q7). After completing all questions, **only upload the finished iPython Notebook file (.ipynb) to Canvas**. Remember to check that you have submitted the file correctly on Canvas (You do not need to submit the files (e.g., csv) you output through your code). **Make sure all cells in your file run without errors after restarting the kernel**.

The deadline to submit your files to Canvas is **Mar. 23rd (Sunday), 11:59 pm**. Please note that this assignment accounts for `20% of your final course score`.

### Background
As you know, the college admissions process involves a lot of types of data from prospective students to make decisions. With the number of applicants increasing, colleges may begin relying on algorithms to select which applications should receive more intensive human review. An algorithm could use quantitative data--such as GPA and SAT score--to provide initial recommendations. In fact, there is more data available than ever. Many colleges even track data about prospective student engagement

> e.g., whether they open emails, visit the college website, engage on social media, etc. This creates a “demonstrated interest” value.

[Based on a survey of college admissions officers](https://www.insidehighered.com/admissions/article/2018/11/12/new-data-admissions-including-application-trends-early-decision-and), we know some of the weights that humans tend to give to these different types of data. Your task will be to create a program that iterates through a list of data points and provides a recommendation for which prospective students are likely to be the best candidates for admission.

Prospective student data is organized in the **admission_algorithms_dataset.csv** file such that the data for each student is on one line, with the values separated by commas. Examples of student data might be:

> Student, SAT, GPA, Interest, High School Quality, Semester 1, Semester 2, Semester 3, Semester 4
>
> Abbess Horror , 1300, 3.61, 10, 7, 95, 86, 91, 94
>
> Adele Hawthorne , 1400, 3.67, 0, 9, 97, 83, 85, 86
>
> Adelicia von Krupp , 900, 4, 5, 2, 88, 92, 83, 72

The data includes these columns (in order):

- **Student**: a unique identifier for each student (their name)
- **SAT score**: value between 0 and 1600
- **GPA**: value between 0 and 5
- **Interest**: value between 0 and 10 (from very low interest to very high interest)
- **High School Quality**: value between 0 and 10 (from very low-quality to very high-quality high school)
- **Semester 1**: average grade for semester 1
- **Semester 2**: average grade for semester 2
- **Semester 3**: average grade for semester 3
- **Semester 4**: average grade for semester 4

## **PART 1 CODING TASKS**

### **Question 1 - Getting set up with our data (15 points)**

First, we need to make sure that we can appropriately read in the data line by line, parsing/converting each line into a list and converting each element in the list to the appropriate type.

### Task 1
Read in your data set and loop through its contents line by line. Make use of the `str.split(delimiter)` function to break each individual line into a list of elements (Name, SAT, GPA, etc.). Ensure that you've done this correctly by printing your list after using the split() function. You'll want to delete this print statement later, but make sure to double check that you are correctly looping over all the lines before moving on! Once you have each line in a list, save the student's name in a variable, then delete this element from your list.

In [7]:
filename = "admission_algorithms_dataset.csv"

input_file = open(filename, "r")

# grab the line with the headers
headers = input_file.readline()

### INSERT YOUR CODE HERE
student_names = []
all_data = []

for line in input_file:
    data = line.strip().split(",")  

    student_name = data.pop(0) 
    student_names.append(student_name)  
    all_data.append(data)  

input_file.close()

### Task 2
Once you have a list of strings for each line, you will write a function `convert_row_type()` that takes one list of elements (representing the data for one student) as a parameter and converts it so that all numbers (currently represented as strings) are converted to floats. Make sure not to lose any information when you do this conversion! Implement this as a pure list function. , i.e., return a new list and do not modify the list passed in.

Example:
> Input: ["1300", "3.61", "10" ,"7" ,"95" ,"86" ,"91" ,"94"]
> 
> Return value: [1300.0, 3.61, 10.0, 7.0, 95.0, 86.0, 91.0, 94.0]

A function `check_row_types()` is provided for you to check your results. Once you've called `convert_row_type()` on the list representing one row, you may call this function to ensure that you have correctly converted all elements to floats.

In [10]:
# Provided code
# This function checks to ensure that a list is of length
# 8 and that each element is type float
# Parameters:
# row - a list to check
# Returns True if the length of row is 8 and all elements are floats
def check_row_types(row):
    if len(row) != 8:
        print("Length incorrect! (should be 8): " + str(row))
        return False
    ind = 0
    while ind < len(row):
        if type(row[ind]) != float:
            print("Type of element incorrect: " + str(row[ind]) + " which is " + str(type(row[ind])))
            return False
        ind += 1
    return True

In [12]:
### INSERT YOUR CODE HERE
def convert_row_type(row):
    return [float(value) for value in row]  

all_converted_data = []

for i in range(len(all_data)): 
    original_data = all_data[i]  
    converted_row = convert_row_type(original_data)
    all_converted_data.append(converted_row) 

    print(original_data) 
    print(converted_row)  
    print(check_row_types(converted_row))
 

['1300', '3.61', '10', '7', '95', '86', '91', '94']
[1300.0, 3.61, 10.0, 7.0, 95.0, 86.0, 91.0, 94.0]
True
['1400', '3.67', '0', '9', '97', '83', '85', '86']
[1400.0, 3.67, 0.0, 9.0, 97.0, 83.0, 85.0, 86.0]
True
['900', '4', '5', '2', '88', '92', '83', '72']
[900.0, 4.0, 5.0, 2.0, 88.0, 92.0, 83.0, 72.0]
True
['1360', '3.19', '0', '9', '90', '85', '97', '87']
[1360.0, 3.19, 0.0, 9.0, 90.0, 85.0, 97.0, 87.0]
True
['1180', '2.93', '2', '7', '93', '81', '90', '87']
[1180.0, 2.93, 2.0, 7.0, 93.0, 81.0, 90.0, 87.0]
True
['1180', '3', '6', '7', '95', '88', '97', '97']
[1180.0, 3.0, 6.0, 7.0, 95.0, 88.0, 97.0, 97.0]
True
['1140', '2.98', '10', '8', '92', '81', '95', '90']
[1140.0, 2.98, 10.0, 8.0, 92.0, 81.0, 95.0, 90.0]
True
['1190', '3.08', '1', '7', '96', '82', '83', '81']
[1190.0, 3.08, 1.0, 7.0, 96.0, 82.0, 83.0, 81.0]
True
['1400', '3.39', '10', '9', '85', '87', '92', '98']
[1400.0, 3.39, 10.0, 9.0, 85.0, 87.0, 92.0, 98.0]
True
['1310', '3.92', '8', '4', '95', '88', '87', '95']
[1310.0,

### Task 3
Separate your data. Use list slicing to separate your list (which should contain 8 numbers at this point) into two lists: one that contains the student's SAT, GPA, Interest, and High School Quality scores, and one that contains their 4 semester grades. You'll do Questions 2 - 4 with the first list of 4 numbers and Question 5 with the list of grades.

In [15]:
### INSERT YOUR CODE HERE
primary_info = []  
semester_grades = []  

for i in range(len(all_converted_data)):  
    primary_info.append(all_converted_data[i][:4])
    semester_grades.append(all_converted_data[i][4:]) 

for i in range(len(primary_info)):
    print(f" Primary Information = {primary_info[i]}, Semester Grades = {semester_grades[i]}")

 Primary Information = [1300.0, 3.61, 10.0, 7.0], Semester Grades = [95.0, 86.0, 91.0, 94.0]
 Primary Information = [1400.0, 3.67, 0.0, 9.0], Semester Grades = [97.0, 83.0, 85.0, 86.0]
 Primary Information = [900.0, 4.0, 5.0, 2.0], Semester Grades = [88.0, 92.0, 83.0, 72.0]
 Primary Information = [1360.0, 3.19, 0.0, 9.0], Semester Grades = [90.0, 85.0, 97.0, 87.0]
 Primary Information = [1180.0, 2.93, 2.0, 7.0], Semester Grades = [93.0, 81.0, 90.0, 87.0]
 Primary Information = [1180.0, 3.0, 6.0, 7.0], Semester Grades = [95.0, 88.0, 97.0, 97.0]
 Primary Information = [1140.0, 2.98, 10.0, 8.0], Semester Grades = [92.0, 81.0, 95.0, 90.0]
 Primary Information = [1190.0, 3.08, 1.0, 7.0], Semester Grades = [96.0, 82.0, 83.0, 81.0]
 Primary Information = [1400.0, 3.39, 10.0, 9.0], Semester Grades = [85.0, 87.0, 92.0, 98.0]
 Primary Information = [1310.0, 3.92, 8.0, 4.0], Semester Grades = [95.0, 88.0, 87.0, 95.0]
 Primary Information = [1310.0, 4.0, 1.0, 1.0], Semester Grades = [90.0, 86.0, 8

### **Question 2 - Prospective Student Score (15 points)**
### Task 1

Write a function `calculate_score()` that takes a list as a parameter and calculates an overall score for each student based on the following weights:

- 30% SAT
- 40% GPA
- 10% demonstrated interest
- 20% strength of curriculum

The list parameter will contain all of the relevant information about the student. The return value is the student’s calculated score. To make this work, you will also need to normalize both GPA and SAT so that they are also on a 0 to 10 scale. To do this, multiply the GPA by 2, and divide the SAT score by 160.

Example:

> Input: [1300.0, 3.61, 10.0, 7.0] - which represents a student with a 1300 SAT score, a 3.61 GPA, 10 out of 10 for interest and 7 out of 10 for high school quality
>
> Output: ((1300 / 160) * 0.3) + ((3.61 * 2) * 0.4) + (10 * 0.1) + (7 * 0.2) = 7.73 out of 10

Round your final output score to 2 decimal places. This can be done following the example:

> `name = "Dr. Stephens"`
> 
> `score = 7.1`
> 
> `print(f"{name},{score:.2f}")`

This will print "Dr. Stephens,7.10"

You start the formatted string with an `f`, then enclose the variables in curly braces {}. The `:.2f` tells the formatter to print the variable as a floating point number with 2 digits after the decimal point (`.`). You can use these formatted strings anywhere you would use a string, i.e. print(), write(), & assignment statements, etc.

In [18]:
### INSERT YOUR CODE HERE
def calculate_score(student_info):
    
    sat = student_info[0] / 160  
    gpa = student_info[1] * 2  
    interest = student_info[2]  
    high_school_quality = student_info[3] 

    score = (sat * 0.3) + (gpa * 0.4) + (interest * 0.1) + (high_school_quality * 0.2)
    
    return round(score, 2)  
    
student_scores = []

for i in range(len(student_names)):  
    student_name = student_names[i] 
    student_score = calculate_score(primary_info[i])
    student_scores.append(student_score)

    print(f"{student_name},{student_score:.2f}") 

Abbess Horror ,7.73
Adele Hawthorne ,7.36
Adelicia von Krupp ,5.79
Adriana Soria ,6.90
Adriana Princess,6.16
Adrienne Frost ,6.61
Agatha Timly ,7.12
Agent Anesthesia ,6.20
Aghanne Buffalo,8.14
Aguja Buffalo,7.19
Akira Kiamata ,5.96
Alaisa Ruatha Pethnan ,6.24
Alchemist Student,7.14
Aldebron James,7.45
Aleksandra Nikolaevna ,6.71
Alflyse Freshman,6.78
Alice Hayes ,6.13
Alice Nugent ,5.73
Alice Tremaine ,6.86
Alizon Baunacha ,7.68
Alkhema Salad,5.51
Allatou Buffalo,6.04
Alma Chalmers ,5.34
Amanda Mueller ,6.29
Amber Alexis ,7.02
Amie Zamborano ,6.72
Amity Hunter ,6.88
Anastasia Kravinoff ,5.80
Anastasia Summit ,7.20
Andrea Haggard ,6.78
Andrea Janson ,7.39
Andrea Roarke ,7.64
Andrea von Strucker ,7.63
Angela Bradford ,7.43
Angela Golden ,7.37
Angelina Brancale ,6.87
Angeline Kutter ,5.53
Anita Ehren ,4.68
Ann Fraley ,6.72
Ann Herd ,5.49
Anna Frankenstein ,5.20
Anne Mutant ,7.55
Anne Marie Cortez ,6.59
Anzhela Federova ,6.65
Aranda Charboneau ,6.61
Araoha Tepania ,7.00
Ariana Saddiqi ,6.1

### Task 2
Call the `calculate_score()` function for each line (row) of data (after you've converted it). Then, write the student's name and their calculated score to a new file called `student_scores.csv` such that each row contains a student’s name and their score, separated by a comma.

Example:

> Abbess Horror , 1300, 3.61, 10, 7, 95, 86, 91, 94
> 
> Adele Hawthorne , 1400, 3.67, 0, 9, 97, 83, 85, 86
> 
> Adelicia von Krupp , 900, 4, 5, 2, 88, 92, 83, 72

lines written to file:

> Abbess Horror , 7.73
> 
> Adele Hawthorne , 7.36
> 
> Adelicia von Krupp , 5.79

In [21]:
### INSERT YOUR CODE HERE
output_filename = "student_scores.csv"

with open(output_filename, "w") as output_file:  
    output_file.write("Student,Score\n") 
    
    for i in range(len(student_names)):  
        student_name = student_names[i]  
        student_score = calculate_score(primary_info[i]) 
        
        output_file.write(f"{student_name},{student_score:.2f}\n")   

Check the `student_scores.csv` file you just create, and print out the answers to the following 3 questions:
- who gets the highest score after calculation, and what is the score?
- who gets the lowest score after calculation, and what is the score?
- what is the average score among all students? Round the results to 2 decimal places.

In [24]:
### INSERT YOUR CODE HERE
input_filename = "student_scores.csv"

students = []
scores = []

with open(input_filename, "r") as file:
    next(file) 
    for line in file:
        name, score = line.strip().split(",")
        score = float(score)  
        students.append(name)
        scores.append(score)

max_index = scores.index(max(scores)) 
min_index = scores.index(min(scores))  
average_score = round(sum(scores) / len(scores), 2)

print(f" Highest Score: {students[max_index]}, {scores[max_index]:.2f}")
print(f" Lowest Score: {students[min_index]}, {scores[min_index]:.2f}")
print(f" Average Score: {average_score:.2f}")



 Highest Score: Betty Sumitro , 8.61
 Lowest Score: Lupe Werewolf, 4.07
 Average Score: 6.65


### Task 3
Write the names for all students who have a score of 6 or higher to a file called `chosen_students.csv`.

Example:

> Abbess Horror , 1300, 3.61, 10, 7, 95, 86, 91, 94
> 
> Adele Hawthorne , 1400, 3.67, 0, 9, 97, 83, 85, 86
> 
> Adelicia von Krupp , 900, 4, 5, 2, 88, 92, 83, 72

lines written to file:

> Abbess Horror
> 
> Adele Hawthorne

Make sure to close all the files you opened in your program.


In [27]:
### INSERT YOUR CODE HERE
input_filename = "student_scores.csv"
output_filename = "chosen_students.csv"

with open(input_filename, "r") as file, open(output_filename, "w") as output_file:
    output_file.write("Student\n") 
    
    next(file)  
    
    for line in file:
        name, score = line.strip().rsplit(",", 1)  
        score = float(score) 
        
        if score >= 6.0: 
            output_file.write(f"{name}\n")


Check the `chosen_students.csv` file you just create, how many students are chosen? Print out the number.

In [30]:
### INSERT YOUR CODE HERE
chosen_filename = "chosen_students.csv"

with open(chosen_filename, "r") as file:
    lines = file.readlines()  

num_students = len(lines) - 1  

print(f" Number of chosen students: {num_students}")


 Number of chosen students: 159


### **Question 3 - Looking for Outliers (15 points)**
Consider ways that this algorithm might systematically miss certain kinds of edge cases. For example, what if a student has a 0 for demonstrated interest because they don’t use social media or have access to a home computer? What if a student has a very high GPA but their SAT score is low enough to bring their score down; could this mean that they had a single bad test taking day?

### Task 1
Write a function `is_outlier()` that can check for certain kinds of outliers. It should check that:
1. if the demonstrated interest score is 0 or
2. if the normalized GPA that is more than 2 points higher than the normalized SAT score.

If either of these conditions is true, it should return `True` (because this student is an outlier); otherwise, the function returns `False`.

In [33]:
### INSERT YOUR CODE HERE
def is_outlier(student_info):

    sat = student_info[0] / 160  
    gpa = student_info[1] * 2  
    interest = student_info[2]  

    if interest == 0 or (gpa - sat > 2):
        return True
    return False



### Task 2
Call `is_outlier()` for each student and write the students' names to a file called `outliers.csv`, one name per line if they are an outlier.

In [36]:
### INSERT YOUR CODE HERE
output_filename = "outliers.csv"

with open(output_filename, "w") as outlier_file:
    outlier_file.write("Student\n") 

    for i in range(len(student_names)):
        student_name = student_names[i]
        if is_outlier(primary_info[i]): 
            outlier_file.write(f"{student_name}\n")  


Check the `outliers.csv` file you just create, how many students are identified as outliers? Print out the number.

In [39]:
### INSERT YOUR CODE HERE
with open("outliers.csv", "r") as file:
    lines = file.readlines()

num_outliers = len(lines) - 1  
print(f" Number of outlier students: {num_outliers}")

 Number of outlier students: 37


### Task 3
Combine the work that you've done now to create an improved list of students to admit to your school. Write students' names, one per line, to the file `chosen_improved.csv` if they either have a score of 6 or greater OR if they are an outlier and their score is 5 or greater. Make sure to take advantage of the work that you’ve already done by calling your functions from previous problems to help you out!

In [42]:
### INSERT YOUR CODE HERE
output_filename = "chosen_improved.csv"

with open(output_filename, "w") as improved_file:
    improved_file.write("Student\n")

    for i in range(len(student_names)):
        student_name = student_names[i]
        student_score = calculate_score(primary_info[i])
        
        if student_score >= 6.0 or (is_outlier(primary_info[i]) and student_score >= 5.0):
            improved_file.write(f"{student_name}\n")

Check the `chosen_improved.csv` file you just create, how many students are chosen by this improved algorithm? Print out the number.

In [45]:
### INSERT YOUR CODE HERE
with open("chosen_improved.csv", "r") as file:
    lines = file.readlines()

num_selected = len(lines) - 1 
print(f" Number of students selected in improved list: {num_selected}")

 Number of students selected in improved list: 171


### **Question 4 - Slightly Improved Algorithm (10 points)**
### Task 1

Create a `calculate_score_improved()` function that calculates a student score, checks if it is an outlier, and returns `True` if the student has a score of 6 or higher OR was flagged as an outlier; otherwise, return `False`. Make sure to take advantage of the work that you've already done by calling your functions from previous problems to help you out!

In [48]:
### INSERT YOUR CODE HERE
def calculate_score_improved(primary_info):
    
    score = calculate_score(primary_info)

    outlier_status = is_outlier(primary_info) 

    return score >= 6 or outlier_status

### Task 2

Call `calculate_score_improved()` and output each student’s information (name, SAT, GPA, interest score, and high school quality) to a new file called `better_improved.csv` if `calculate_score_improved()` returned `True` for them.

In [51]:
### INSERT YOUR CODE HERE
output_filename = "better_improved.csv"

with open(output_filename, "w") as file:
    file.write("Student,SAT,GPA,Interest,High School Quality\n")

    for i in range(len(student_names)):
        student_name = student_names[i]  
        student_primary = primary_info[i]  

        if calculate_score_improved(student_primary):
            file.write(f"{student_name},{student_primary[0]},{student_primary[1]},{student_primary[2]},{student_primary[3]}\n")


Check the `better_improved.csv` file you just create, how many students are chosen by this new algorithm? Print out the number.

In [54]:
### INSERT YOUR CODE HERE
with open("better_improved.csv", "r") as file:
    lines = file.readlines()

num_selected_students = len(lines) - 1 

print(f" Total students selected by the improved algorithm: {num_selected_students}")

 Total students selected by the improved algorithm: 174


### **Question 5 - GPA Checker (15 points)**

A single GPA score is not a full picture of a student’s academic performance, as it may have improved over time or included outlier courses or semesters. A more context-sensitive algorithm could consider a student’s entire transcript and checks for, for example, a single class score that is more than two letter grades (20 points) lower than all other scores. For this task, you will use the second half of the data for each student in the provided file.

### Task 1
Write a function `grade_outlier()` that takes in a list of grades (of any length) and returns `True` if one single number is more than 20 points lower than all other numbers; otherwise, `False`.

Example:

> Input: [99, 94, 87, 89, 56, 78, 89]

Hint: Sort the list from lowest to highest, and check for the difference between the two lowest grades.

78 - 56 = 22; 22 > 20

> Output: True

Next, consider the data that we have: a list of grades for each student, one grade per semester for four semesters.

Make sure that your `grade_outlier()` function works by calling it for every row in the second dataset. Print out an informative message about which students have a single grade outlier. You'll delete this later but it's a great way of testing your function!

Finally, consider the importance of an algorithm being able to flag students who might have a lower overall GPA but have shown improvement over time.

In [57]:
### INSERT YOUR CODE HERE
def grade_outlier(grades):
    grades = sorted(grades)  
    
    if len(grades) < 2: 
        return False
    
    if grades[1] - grades[0] > 20:  
        return True 
    
    return False 

### Task 2
Create a function `grade_improvement()` that returns `True` if the average score of each semester is higher than or equal to each previous semester and `False` otherwise.

Hint: investigate how the `==` operator works between two lists and think about using the Python's `sorted()` function.

In [60]:
### INSERT YOUR CODE HERE

def grade_improvement(grades):
    return grades == sorted(grades)  

### Task 3

Using the grade information that you've just learned, create conditions based on the information from the previous problems and `grade_outlier()` and `grade_improvement()` to choose all students if they either have a score of 6 or greater or if they have a score of 5 or more and at least one of the following is true:

- `is_outlier()` returns `True`
- `grade_outlier()` returns `True`
- or `grade_improvement()` returns `True`

Write the students who fit this description to `composite_chosen.csv`, one name per line.


In [63]:
### INSERT YOUR CODE HERE
output_filename = "composite_chosen.csv"

with open(output_filename, "w") as file:
    file.write("Student\n") 

    for i in range(len(student_names)):
        student_name = student_names[i] 
        student_score = student_scores[i] 
        student_grades = semester_grades[i]
        student_primary = primary_info[i] 

        if (student_score >= 6) or \
           (student_score >= 5 and (is_outlier(student_primary) or grade_outlier(student_grades) or grade_improvement(student_grades))):

            file.write(f"{student_name}\n")


Check the `composite_chosen.csv` file you just create, how many students are chosen by this algorithm? Print out the number.

In [66]:
### INSERT YOUR CODE HERE
with open("composite_chosen.csv", "r") as file:
    lines = file.readlines()

num_selected_students = len(lines) - 1  

print(f" Total students selected in composite_chosen.csv: {num_selected_students}")


 Total students selected in composite_chosen.csv: 174


**Note**: 

We will be testing your program using a file with the same format but different data! Make sure that you haven't "hard coded" anything specific to your data.

## **PART 2 REFLECTION TASKS**

Based on the algorithm you just implemented, answer the following questions:

### **Question 6 - Reflect on the outcomes (15 points)**

In 100 to 200 words, reflect on the outcomes of your implemented algorithms. What are some examples of ways these algorithms might still give undesirable results for certain types of prospective students?

Type your answer here:


- There are some issues with setting outliers.  

1. For example, interest equals to 0 cannot be absolute outlier. Not all of them don't use social media or have access to a home computer. It may also be due to their procrastination or lack of information gathering ability. Even without a computer and social media, they can still show their interests through sending letters, participating in offline activities organized by the school, making phone calls, and others to show their interests 

2. In my opinion, when the grades are abnormally low in a certain semester, it is also important to calculate the grades of the other three semesters. If one grade is extremely low, influencing the other three high grades(like 10, 90, 90, 90), then leads to not meet the requirements, it will miss some capable people. At the same time, a abnormal grade in a certain semester cannot be completely classified as an outlier. If due to family reasons such as father’s death, or getting seriously ill can indeed be considered as an outlier. but if is due to stress, that grade can't be regarded as outlier. Since self-regulation ability in the face of pressure is also important ability 

### **Question 7 - How should and shouldn’t algorithms be used in college admissions? Why? (15 points)**

Given the realities of time and resources (i.e., only limited person-hours can be involved in the admission process), do you think we should use algorithms like these for the admission process (assuming the real-world algorithm is far more sophisticated than the one you just created)? Explain your thinking, taking into account possible undesirable outcomes.

Type your answer here:

- During the admission process, universities need to process a large amount of data received. If these data are processed and filtered by human to determine which students meet the requirements, it is very easy to make mistakes and biases and will consume a lot of time and resources. 

    The use of algorithms can efficiently process a large number of applications, quickly select the applicants who do not meet the requirements, reducing human errors and biases, and improve accuracy. For example, students with GPA higher than a certain value can be quickly selected, while students with low overall levels can be quickly excluded. At the same time, it will not consume huge human resources. 

    But no matter how advanced the algorithm is, it is always based on existing data to have training and then make decisions. If there are problems with the training data, the algorithm will be biased, which will affect the fairness. 
    At the same time, for those students with low-come and less resources, they may be at a disadvantage in exam, extracurricular activities and so on. Even if they have strong learning potential, algorithms may not prioritize recommending them, but instead choose students from affluent families because their schools have richer resources, can provide better SAT preparation guidance, and have more extracurricular activity resources 
    Algorithms can deal with standardized data such as grades very easy, but it is not so good when handle personal experiences, extracurricular activities, social responsibility, and other such data, which may lead to misjudging the abilities of certain students 

    In summary, my suggestion is to adopt an algorithm+manual review model. Algorithms have great advantages in handling standardized data (SAT, GPA, grade), while manual review can compensate for the shortcomings of algorithms in handling student background, personal experience, social responsibility, and other data.  

    So I think algorithms need to be used, in order to improve efficiency and accuracy. But at the same time, manual review must also be employed, further improve accuracy, and compensate for the disadvantages of the algorithm I mentioned earlier. 